### ISSUE TITLE
Proposal - Add Service/Plan instances to Context Object

### ISSUE THREAD

**Issue by kisamoto** [pivotal-cf/brokerapi#62]
# What?

In relevant HTTP handlers add checked service/plan instances to HTTP context for further reuse throughout the platform.

# Why?

The API implemented in the brokerapi library is already retrieving the list of services and performing various checks (serviceID & planID valid etc.) however this data is just discarded and methods looking to reuse the service instance have to retrieve the service instance again. 

This inefficiency can be removed by adding the valid service instance to the `context.Context` of the request being processed.

# How?

_Code untested_

Example using [`brokerapi/api.go:serviceBrokerHandler.provision`](https://github.com/pivotal-cf/brokerapi/blob/master/api.go#L141)

Add new methods for adding & extracting Service and Plan objects from `Context`: 

```golang
type contextKey string

const (
	contextKeyService contextKey = "brokerapi_service"
	contextKeyPlan    contextKey = "brokerapi_plan"
)

func AddServiceToContext(service *Service, ctx context.Context) context.Context {
	return context.WithValue(ctx, contextKeyService, service)
}

func AddPlanToContext(plan *Plan, ctx context.Context) context.Context {
	return context.WithValue(ctx, contextKeyPlan, plan)
}

func ServiceFromContext(ctx context.Context) *Service {
	if v := ctx.Value(contextKeyService); v != nil {
		return v.(*Service)
	}
	return nil
}

func PlanFromContext(ctx, context.Context) *Plan {
	if v := ctx.Value(contextKeyPlan); v != nil {
		return v.(*Plan)
	}
	return nil
}
```

Update api to add details to `Context`

```golang
[..]
	services, _ := h.serviceBroker.Services(req.Context())
	for _, service := range services {
		if service.ID == details.ServiceID {
			req = req.WithContext(AddServiceToContext(*service, req.Context())) // Add to Context
			valid = true
			break
		}
	}
[..]
	for _, service := range services {
		for _, plan := range service.Plans {
			if plan.ID == details.PlanID {
				req = req.WithContext(AddPlanToContext(*plan, req.Context())) // Add to Context
				valid = true
				break
			}
		}
	}
[..]
```

Easily re-use these instances from `Context` in ServiceBroker implementation

```golang
service := brokerapi.ServiceFromContext(ctx)
if service == nil {
        // fallback to looking up service from database
}
```

# Next steps

Open to discussion about pros/cons of this implementation. If it's something that may be of interest I can write a complete pull request.

**Comment by jacknewberry** [comment]
Hi Ewan,
Looks good to us. Seems like an appropriate use of the context object.

We would be very happy to receive a complete PR.

Jack & @kirederik

**Comment by FelisiaM** [comment]
Change is merged as part of #63.
